Add storybook and a Header story

marcaaron opened the pull request:
### Details
Adds Storybook to E.cash

### Fixed Issues
Fixes https://github.com/Expensify/Expensify/issues/164021

### Tests
1. `npm i`
2. Run `npm run storybook`
3. Verify that storybook opens

### QA
1. Navigate to `https://staging.expensify.cash/docs`
1. Verify that storybook is accessible

### Tested On

- [X] Web
-  Mobile Web
-  Desktop
-  iOS
-  Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
![2021-05-13_14-04-52](https://user-images.githubusercontent.com/32969087/118201868-35be7b00-b3f4-11eb-85fe-0c110f299384.png)

marcaaron committed:
Add storybook and a Header story

marcaaron committed:
fix conflicts

marcaaron committed:
make as few changes as possible

botify commented:
Hey, I noticed you changed some webpack configuration files. This can break production builds. Did you remember to run a production build locally to verify they still work?

marcaaron committed:
clean up

marcaaron committed:
fix conflicts

marcaaron committed:
Fix conflicts

marcaaron committed:
Use staging config

botify commented:
Hey, I noticed you changed some webpack configuration files. This can break production builds. Did you remember to run a production build locally to verify they still work?

botify commented:
Hey, I noticed you changed some webpack configuration files. This can break production builds. Did you remember to run a production build locally to verify they still work?

roryabraham commented on babel.config.js:
```diff
@@ -32,6 +32,6 @@ const metro = {
     // For `react-native` (iOS/Android) caller will be "metro"
     // For `webpack` (Web) caller will be "@babel-loader"
-    const runningIn = caller(({name}) => name);
+    const runningIn = caller((args = {}) => args.name);
```
NAB but why was this necessary?

marcaaron replied:
When storybook loads the babel config there is no `args` object so some things were quietly breaking. Specifically, parsing the docs out of `propTypes` would not work. Maybe I should add some comments so nobody breaks this.

marcaaron replied:
Added some comment here about the significance to SB

roryabraham commented on src/stories/Header.stories.js:
```diff
@@ -0,0 +1,17 @@
+import Header from '../components/Header';
+
+export default {
```
Can you explain what's going on in this file? Maybe a link to some docs? Should we add some instructions to the `STYLE.md` style guide for creating `*.stories.js` files?

marcaaron replied:
Up for that, but might choose to do it in a follow up. There's a lot of documentation about Storybook available so I'm not sure what would go the style guide at this point except maybe something like "head to the project site and learn more about how to create stories".

marcaaron replied:
Added some links to the docs. Could do something more substantial eventually.

roryabraham reviewed (commented):
GH actions code seems fine!

marcaaron committed:
Add comment

marcaaron committed:
add some comments

marcaaron committed:
use better names

shawnborton commented:
This is probably not the best place to deliver this feedback, but what is the difference between default and large for the header text size? I actually think large (assuming it's 17) is probably the default here. In fact, I think the only other time we would have a non-17 sized font in the header area is for chats. And I think that's okay - it's almost like we aren't using a font-size for the header title but rather, we are using a component as the header title - with the component being the little avatar(single, two, or counter) + name(s) component we use.

marcaaron commented:
Oh hmm yeah that's a good point @shawnborton. Looking through the code now and I don't really see anything passing this prop at all. I think all of the headers are set to have "large" text.

marcaaron commented:
Actually this is only true for `HeaderWithCloseButton`

https://github.com/Expensify/Expensify.cash/blob/66d265e144a69f0a40bf5ca05d1b506a5a19a4bd/src/components/HeaderWithCloseButton.js#L37-L45

marcaaron commented:
e.g. the IOU view is not using the large text

<img width="376" alt="Screen Shot 2021-05-14 at 7 33 40 AM" src="https://user-images.githubusercontent.com/32969087/118307684-bcbc3380-b486-11eb-9d1d-30f1b3728763.png">
<img width="376" alt="Screen Shot 2021-05-14 at 7 33 31 AM" src="https://user-images.githubusercontent.com/32969087/118307688-bded6080-b486-11eb-804e-99eb595e04e9.png">

shawnborton commented:
Ah got it, the IOU view should be using the larger text size. I say we remove it as a prop and just update the default header styles everywhere to use 17px as the font size unless a component replaces the header title.

marcaaron commented:
We could probably change this to `textSize="small"` and have the default be `17` if that makes more sense.

marcaaron committed:
remove textSize from Header

roryabraham reviewed (dismissed):
Just tested locally and this is **AWESOME!!!** I'm stoked for this – is there a tracking issue to create stories for all our components? 😁

marcaaron committed:
Fix up the HeaderWithCloseButton

marcaaron committed:
add a few stories

marcaaron commented:
Fixed up that Header issue @shawnborton and added some stories for the more complex `HeaderWithCloseButton`. It shows off the variations:

https://user-images.githubusercontent.com/32969087/118315456-407b1d80-b491-11eb-8c85-20d3559f1357.mp4

marcaaron committed:
fix conflicts;

shawnborton commented:
Woo, thanks Marc!

roryabraham commented on .storybook/webpack.config.js:
```diff
@@ -0,0 +1,35 @@
+    config.resolve.alias = {
+        'react-native-config': 'react-web-config',
+        'react-native$': 'react-native-web',
+        '@react-native-community/netinfo': './mock',
```
Sorry if I missed this in a previous review, but can we just do this?

```suggestion
        '@react-native-community/netinfo': path.resolve(__dirname, '../__mocks__/@react-native-community/netinfo.js'),
```

Then you don't need to create `src/libs/NetInfo/mock.js`

marcaaron replied:
Won't really work because we are not running in `jest` and that mock depends on it. We could maybe not use the default mock tho with the jest.fn() lemme try and get back to ya.

marcaaron replied:
Alright this worked after getting rid of the jest dependency so I got rid of that other file :) Nice one!

thienlnam commented:
Wowwww this is so cool

![Screen Shot 2021-05-14 at 1 04 14 PM](https://user-images.githubusercontent.com/30609178/118323056-ed1ac680-b4b4-11eb-89e6-62ae7e4d3017.png)

marcaaron committed:
remove unneeded mock but create new jest free version of mock

roryabraham reviewed (approved):
Cool!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/Expensify.cash/actions/runs/843695847) to staging in version: 1.0.46-1🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

isagoico commented:
@roryabraham Heyo! Are we able to test this? I tried navigating to https://staging.expensify.cash/docs and I was unable to access storybook.

marcaaron commented:
@isagoico can check this off as we'll need to fix this somewhere else but can access via `/docs/index.html` for now. It's not a blocker in any case.
